packages/lodestar/src/chain/chain.ts

```diff
@@ -193,6 +202,88 @@ export class BeaconChain extends (EventEmitter as { new(): ChainEventEmitter })
     };
   }
 
+  public async waitForBlockProcessed(blockRoot: Uint8Array): Promise<void> {
+    await new Promise((resolve) => {
+      this.on("processedBlock", (signedBlock) => {
+        const root = this.config.types.BeaconBlock.hashTreeRoot(signedBlock.message);
+        if (this.config.types.Root.equals(root, blockRoot)) {
+          resolve();
+        }
+      });
+    });
+  }
+
+  /**
+   * Restore state cache and forkchoice from last finalized state.
+   */
+  private async restoreHeadState(lastKnownState: TreeBacked<BeaconState>): Promise<void> {
+    this.logger.info(`Found last known/finalized state at epoch #${lastKnownState.finalizedCheckpoint.epoch}`);
+    this.logger.profile("restoreHeadState");
+    this.db.stateCache.add(lastKnownState);
+    // the block respective to finalized epoch is still in block db
+    const allBlocks = await this.db.block.values();
+    if (!allBlocks || allBlocks.length === 0) {
+      return;
+    }
+    const sortedBlocks = sortBlocks(allBlocks);
+    const firstBlock = allBlocks[0];
+    const lastBlock = allBlocks[allBlocks.length - 1];
+    let firstSLot = firstBlock.message.slot;
+    let lastSlot = lastBlock.message.slot;
+    this.logger.info(`Found ${allBlocks.length} blocks in database, from slot ${firstSLot} to ${lastSlot}`);
+    if (allBlocks.length === 1) {
```

Comment:
=== 0?

Reply:
initially we stored the genesis block to database already so this is to catch that case

Reply:
can you add comment about that as it's hard to figure out